homeassistant/components/enphase_envoy/diagnostics.py

```diff
@@ -36,11 +41,80 @@ async def async_get_config_entry_diagnostics(
     """Return diagnostics for a config entry."""
     coordinator: EnphaseUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
 
-    return async_redact_data(
-        {
-            "entry": entry.as_dict(),
-            "envoy_firmware": coordinator.envoy.firmware,
-            "data": coordinator.data,
-        },
-        TO_REDACT,
+    if TYPE_CHECKING:
+        assert coordinator.envoy.data
+    envoy_data: EnvoyData = coordinator.envoy.data
+    envoy: Envoy = coordinator.envoy
+
+    device_registry = dr.async_get(hass)
+    entity_registry = er.async_get(hass)
+
+    devices = []
+    # for each device associated with the envoy get entity and state information
+    for device in dr.async_entries_for_config_entry(device_registry, entry.entry_id):
+        entities = []
+        for entity in er.async_entries_for_device(
+            entity_registry, device_id=device.id, include_disabled_entities=True
+        ):
+            state_dict = None
+            if state := hass.states.get(entity.entity_id):
+                state_dict = dict(state.as_dict())
+                state_dict.pop("context", None)
+            entities.append({"entity": asdict(entity), "state": state_dict})
+        devices.append({"device": asdict(device), "entities": entities})
+
+    # redact envoy serial in entity data
+    device_entities = json.loads(
+        json.dumps(devices).replace(coordinator.envoy_serial_number, "<<envoyserial>>")
+    )
+
+    # redact envoy serial in envoy data
+    cleaned_data = json.loads(
+        json.dumps(coordinator.data).replace(
+            coordinator.envoy_serial_number, "<<envoyserial>>"
+        )
     )
+
+    envoy_model: dict[str, Any] = {
+        "encharge_inventory": envoy_data.encharge_inventory,
+        "encharge_power": envoy_data.encharge_power,
+        "encharge_aggregate": envoy_data.encharge_aggregate,
+        "enpower": envoy_data.enpower,
+        "system_consumption": envoy_data.system_consumption,
+        "system_production": envoy_data.system_production,
+        "system_consumption_phases": envoy_data.system_consumption_phases,
+        "system_production_phases": envoy_data.system_production_phases,
+        "ctmeter_production": envoy_data.ctmeter_production,
+        "ctmeter_consumption": envoy_data.ctmeter_consumption,
+        "ctmeter_production_phases": envoy_data.ctmeter_production_phases,
+        "ctmeter_consumption_phases": envoy_data.ctmeter_consumption_phases,
+        "dry_contact_status": envoy_data.dry_contact_status,
+        "dry_contact_settings": envoy_data.dry_contact_settings,
+        "inverters": envoy_data.inverters,
+        "tariff": envoy_data.tariff,
+    }
+
+    envoy_properties: dict[str, Any] = {
+        "envoy_firmware": str(envoy.firmware),
```

Comment:
Why the need to cast to str?

Reply:
Was to deal with some test snapshot issues initially. Verified again and removed. Thanks.